source/Mlos.Python/mlos/Grpc/unit_tests/TestOptimizerServiceEncoderDecoder.py

```diff
@@ -103,10 +106,86 @@ def test_composite_dimension(self):
 
     def test_hypergrid(self):
         parameter_space = bayesian_optimizer_config_store.parameter_space
-        serialized = OptimizerServiceEncoder.encode_simple_hypergrid(parameter_space)
-        deserialized = OptimizerServiceDecoder.decode_simple_hypergrid(serialized)
+        serialized = OptimizerServiceEncoder.encode_hypergrid(parameter_space)
+        deserialized = OptimizerServiceDecoder.decode_hypergrid(serialized)
 
         print(deserialized)
 
         for _ in range(1000):
             assert deserialized.random() in parameter_space
+
+        for _ in range(1000):
+            assert parameter_space.random() in deserialized
+
+    def test_optimization_problem(self):
+        parameter_space = SimpleHypergrid(
+            name="test",
+            dimensions=[
+                ContinuousDimension(name="x",min=0,max=1),
+                CategoricalDimension(name="y",values=[1,2,3])
+            ]
+        )
+        objective_space = SimpleHypergrid(
+            name="z",
+            dimensions=[
+                ContinuousDimension(name="z",min=0,max=1),
+                ContinuousDimension(name="z1",min=-1,max=1)
+            ]
+        )
+        context_space = SimpleHypergrid(
+            name="context_space",
+            dimensions=[
+                ContinuousDimension(name="x_c",min=0,max=1),
+                CategoricalDimension(name="y_c",values=[1,2,3,4,6])
+            ]
+        )
+        optimization_problem = OptimizationProblem(
+            parameter_space=parameter_space,
+            objective_space=objective_space,
+            objectives=[
+                Objective(name="z",minimize=True),
+                Objective(name="z1",minimize=False)
+            ],
+            context_space=context_space
+        )
+        encoded_problem = OptimizerServiceEncoder.encode_optimization_problem(optimization_problem)
+        decoded_problem = OptimizerServiceDecoder.decode_optimization_problem(encoded_problem)
+
+        # A = B iff A >= B && B <= A
+        # Could be condensed to single loop but easier to read this way.
+        # Parameter Space
+        for _ in range(1000):
+            assert decoded_problem.parameter_space.random() in parameter_space
+            assert parameter_space.random() in decoded_problem.parameter_space
+
+        # Output Space
+        for _ in range(1000):
+            assert decoded_problem.objective_space.random() in objective_space
+            assert objective_space.random() in decoded_problem.objective_space
+
+        # Context Space
+        for _ in range(1000):
+            assert decoded_problem.context_space.random() in context_space

```

Comment:
Add a similar test for optimization_problem.feature_space

Reply:
Also, please copy this into the `test_optimization_problem_none_context` so that we make sure that the feature space is behaving properly even when the context is set to None

Reply:
[zaporter/grpc 31cc56f5] Adam's suggestions